frappe/model/db_query.py

```diff
@@ -740,8 +740,12 @@ def prepare_filter_condition(self, f):
 		df = meta.get("fields", {"fieldname": f.fieldname})
 		df = df[0] if df else None
 
-		# primary key is never nullable, modified is usually indexed by default and always present
-		can_be_null = f.fieldname not in ("name", "modified")
+		# mandatory fields are never null
+		can_be_null = not getattr(df, "reqd", False)
```

Comment:
Let's not assume this? At db level they can technically be null.

Instead we should tick not null when mandatory is checked so db schema is updated